File: java/src/test/java/com/secacon/secasign/sdk/document/DocumentTest.java
package com.secacon.secasign.sdk.document;

import com.secacon.secasign.extension.SdkConfiguration;
import com.secacon.secasign.extension.SdkTest;
import com.secacon.secasign.extension.SdkTestRequirements;
import com.secacon.secasign.sdk.SecasignHttpClient;
import com.secacon.secasign.sdk.dto.authentication.AuthenticationDto;
import com.secacon.secasign.sdk.dto.authentication.TokenDto;
import com.secacon.secasign.sdk.dto.document.*;
import org.junit.jupiter.api.Assertions;

import java.io.File;
import java.nio.file.Files;
import java.util.Collections;
import java.util.List;
import java.util.UUID;

import static org.junit.jupiter.api.Assertions.assertEquals;

/**
 * Test for showing document information, searching and downloading.
 *
 * @author Simon Wächter
 */
@SdkTest
public class DocumentTest {

    /**
     * Show the document information.
     *
     * @param sdkConfiguration SDK configuration
     * @throws Exception Exception in case of a problem
     */
    @SdkTestRequirements(isDocumentAvailable = true)
    public void testDocumentInformation(SdkConfiguration sdkConfiguration) throws Exception {
        // Create the HTTP client
        SecasignHttpClient secasignHttpClient = new SecasignHttpClient(sdkConfiguration.getUrl());

        // Obtain the token
        TokenDto tokenDto = secasignHttpClient.login(new AuthenticationDto(sdkConfiguration.getEmailAddress(), sdkConfiguration.getPassword()));

        // Get the document information by the document ID
        ReadDocumentDto readDocumentDto = secasignHttpClient.getDocumentById(tokenDto, sdkConfiguration.getDocumentId());
        assertEquals(sdkConfiguration.getDocumentId(), readDocumentDto.getId());
        assertEquals(SignatureTypeDto.PDF, readDocumentDto.getSignatureType());
    }

    /**
     * Search for documents and download them.
     *
     * @param sdkConfiguration SDK configuration
     * @throws Exception Exception in case of a problem
     */
    @SdkTestRequirements(isArchiveSearchAvailable = true)
    public void testDocumentSearchAndDownload(SdkConfiguration sdkConfiguration) throws Exception {
        // Create the HTTP client
        SecasignHttpClient secasignHttpClient = new SecasignHttpClient(sdkConfiguration.getUrl());

        // Obtain the token
        TokenDto tokenDto = secasignHttpClient.login(new AuthenticationDto(sdkConfiguration.getEmailAddress(), sdkConfiguration.getPassword()));

        // Search for the latest documents: First page with index 0, at most 10 document per page and by archive attribute with value
        List<ReadDocumentAttributeDto> readDocumentAttributeDtos = Collections.singletonList(new ReadDocumentAttributeDto(sdkConfiguration.getArchiveAttributeId(), sdkConfiguration.getArchiveAttributeValue()));
        DocumentSearchCriteriaDto documentSearchCriteriaDto = new DocumentSearchCriteriaDto(0, 10, null, null, null, null, null, null, null, null, null, null, null, readDocumentAttributeDtos);
        DocumentSearchResultDto documentSearchResultDto = secasignHttpClient.searchDocuments(tokenDto, documentSearchCriteriaDto);

        // Fail if no documents were found
        if (documentSearchResultDto.getTotalDocuments() == 0) {
            Assertions.fail("No documents were found.");
        }

        // Get the first document to get its document ID
        ReadDocumentDto readDocumentDto = documentSearchResultDto.getDocuments().get(0);
        assertEquals(SignatureTypeDto.PDF, readDocumentDto.getSignatureType());
        UUID documentId = readDocumentDto.getId();

        // Get the current timestamp
        long timestamp = System.currentTimeMillis();

        // Download the original PDF document
        byte[] originalPdfDocumentData = secasignHttpClient.downloadOriginalDocument(tokenDto, documentId);
        File originalPdfDocument = new File("../data/Downloaded_Original_" + timestamp + ".pdf");
        Files.write(originalPdfDocument.toPath(), originalPdfDocumentData);
        System.out.println("The original document was successfully downloaded. Storing it at " + originalPdfDocument.getAbsolutePath());

        // Download the signed PDF document
        byte[] signedPdfDocumentData = secasignHttpClient.downloadSignedDocument(tokenDto, documentId);
        File signedPdfDocument = new File("../data/Downloaded_Signed_" + timestamp + ".pdf");
        Files.write(signedPdfDocument.toPath(), signedPdfDocumentData);
        System.out.println("The signed document was successfully downloaded. Storing it at " + signedPdfDocument.getAbsolutePath());
    }
}

File: csharp/SecasignTest/sdk/document/DocumentTest.cs
using Secasign;
using Secasign.dto.authentication;
using Secasign.dto.document;
using SecasignTest.extension;

namespace SecasignTest.sdk.document;

public class DocumentTest
{
    private static SdkConfiguration SdkConfiguration => SdkExtension.SdkConfiguration;

    [SdkExtension(SdkTestRequirements.IsDocumentAvailable)]
    public async Task TestDocumentInformation()
    {
        // Create the HTTP client
        var secasignHttpClient = new SecasignHttpClient(SdkConfiguration.Url!);

        // Login and obtain a JWT token
        var authenticationDto = new AuthenticationDto(SdkConfiguration.EmailAddress!, SdkConfiguration.Password!);
        var tokenDto = await secasignHttpClient.Login(authenticationDto);

        // Get the document ID
        var documentId = (Guid)SdkConfiguration.DocumentId!;

        // Get the document information by the document ID
        var readDocumentDto = await secasignHttpClient.GetDocumentById(tokenDto, documentId);
        Assert.Equal(SdkConfiguration.DocumentId, readDocumentDto.id);
        Assert.Equal(SignatureTypeDto.PDF, readDocumentDto.signatureType);
    }

    [SdkExtension(SdkTestRequirements.IsArchiveSearchAvailable)]
    public async Task TestDocumentSearchAndDownload()
    {
        // Create the HTTP client
        var secasignHttpClient = new SecasignHttpClient(SdkConfiguration.Url!);

        // Login and obtain a JWT token
        var authenticationDto = new AuthenticationDto(SdkConfiguration.EmailAddress!, SdkConfiguration.Password!);
        var tokenDto = await secasignHttpClient.Login(authenticationDto);

        // Get the archive attribute ID
        var archiveAttributeId = (Guid)SdkConfiguration.ArchiveAttributeId!;

        // Search for the latest documents: First page with index 0, at most 10 document per page and by archive attribute with value
        List<ReadDocumentAttributeDto> readDocumentAttributeDtos = [new(archiveAttributeId, SdkConfiguration.ArchiveAttributeValue!)];
        var documentSearchCriteriaDto = new DocumentSearchCriteriaDto(0, 10, null, null, null, null, null, null, null, null, null, null, null, readDocumentAttributeDtos);
        var documentSearchResultDto = await secasignHttpClient.SearchDocuments(tokenDto, documentSearchCriteriaDto);

        // Fail if no documents were found
        if (documentSearchResultDto.totalDocuments == 0)
        {
            Assert.Fail("No documents were found.");
        }

        // Get the first document to get its document ID
        var readDocumentDto = documentSearchResultDto.documents.First();
        Assert.Equal(SignatureTypeDto.PDF, readDocumentDto.signatureType);
        var documentId = readDocumentDto.id;

        // Get the current timestamp
        var timestamp = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();

        // Download the original PDF document
        var originalPdfDocumentData = await secasignHttpClient.DownloadOriginalDocument(tokenDto, documentId);
        var originalPdfDocument = new FileInfo("../../../../../data/Downloaded_Original_" + timestamp + ".pdf");
        await File.WriteAllBytesAsync(originalPdfDocument.FullName, originalPdfDocumentData);

        // Download the signed PDF document
        var signedPdfDocumentData = await secasignHttpClient.DownloadSignedDocument(tokenDto, documentId);
        var signedPdfDocument = new FileInfo("../../../../../data/Downloaded_Signed_" + timestamp + ".pdf");
        await File.WriteAllBytesAsync(signedPdfDocument.FullName, signedPdfDocumentData);
    }
}
